### PR TITLE
feat: consider squashed call signatures of function components

### PR DESCRIPTION
**Pull request by eps1lon**
input:
```ts
interface Component {
	(props: ButtonProps): JSX.Element;
	(props: { component: React.ElementType } & ButtonProps): JSX.Element;
}
```
output:
```js
ButtonBase.propTypes = {
	component: PropTypes.elementType,
	variant: PropTypes.string,
};
```

TODO:
- [x] props that are only required in some call signatures must be optional

### CODE REVIEW

**Comment by eps1lon** [comment]
@merceyz Anything wrong with this approach?

**Comment by merceyz** [comment]
@eps1lon First off, thanks for the PRs, really appreciate it :+1:. I haven't had the time to properly review this yet, just merged in master to get the tests to run on PRs again. I'll take a look as soon as I have time.

File: test/overloaded-function-component/output.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+ButtonBase.propTypes = {
+	component: PropTypes.oneOf(['div']),
```
Comment (merceyz) [inline test/overloaded-function-component/output.js]: This doesn't seem right, I guess typescript doesn't give you more details because of the default?

Comment (eps1lon) [reply test/overloaded-function-component/output.js]: Yeah. I think this is unrelated though. I'll verify

Comment (eps1lon) [reply test/overloaded-function-component/output.js]: Ok on master they're not working at all. This is better but we need to special case generics to get the widest type.

Comment (eps1lon) [reply test/overloaded-function-component/output.js]: @merceyz That'll do it: [https://github.com/merceyz/typescript-to-proptypes/pull/20/commits/7b900ce4ee4067ed9721e12a130f4573915bd12c ](https://github.com/merceyz/typescript-to-proptypes/pull/20/commits/7b900ce4ee4067ed9721e12a130f4573915bd12c)

Comment (merceyz) [reply test/overloaded-function-component/output.js]: Perfect, thank you

**Review by merceyz — APPROVED** [review]
Perfect, thank you for implementing this and sorry for taking so long

**Comment by merceyz** [comment]
Released as `typescript-to-proptypes@2.0.0`

**Pull request merged**